### ISSUE TITLE
Pyright does not correctly handle clearing __hash__ on subclasses which implement __eq__

### ISSUE THREAD

**Issue by bijij** [microsoft/pyright#3223]
**Describe the bug**
A quote from the [python docs](https://docs.python.org/3/reference/datamodel.html#object.__hash__):
> A class that overrides `__eq__()` and does not define `__hash__()` will have its `__hash__()` implicitly set to `None`. 

> If a class that overrides `__eq__()` needs to retain the implementation of `__hash__()` from a parent class, the interpreter must be told this explicitly by setting `__hash__ = <ParentClass>.__hash__`.

**To Reproduce**
See the sample code below, pyright returns the following.
```
test.py:14:13 - information: Type of "Bar.__hash__" is "(self: Bar) -> int"
```

**Expected behavior**
At runtime python 3.11 prints the following:
```
Runtime type is 'NoneType'
```

**Screenshots or Code**
```python
from typing import reveal_type

class Foo:

    def __hash__(self) -> int:
        return 0


class Bar(Foo):
    def __eq__(self, other: object):
        return False


reveal_type(Bar.__hash__)

```

**VS Code extension or command-line**
pyright 1.1.231

**Comment by erictraut** [comment]
This is pretty esoteric. I don't think it's worth adding a special case to pyright to model this behavior. Mypy doesn't either. If you are relying on this implicit behavior in your class _and_ it matters for type checking purposes, you could explicitly set `__hash__` to `None`.

**Comment by bijij** [comment]
I wasn't actually relying on this behaviour, I prior to this didn't know it existed. 
I instead had a class which implemented `__hash__`, and made a subclass with a custom `__eq__`.

I then tried to use an instance of this class in a Set, which type-checked fine, but failed at runtime has the subclass was unhashable. The type-checker returning that `__hash__` existed left me quite perplexed.

Not sure if `__hash__` being set to `None` would have helped here much though, seeing as `dict` and `set`'s key type generic isn't bound to `collections.abc.Hashable` anyway.
